server/src/main/java/org/elasticsearch/common/bytes/PagedBytesReference.java

```diff
@@ -48,6 +49,26 @@ public BytesReference slice(int from, int length) {
         return new PagedBytesReference(byteArray, offset + from, length);
     }
 
+    @Override
+    public BytesReference deepCopy(int from, int length) throws IOException {
+        Objects.checkFromIndexSize(from, length, this.length);
+        final int offsetFirstPage = (Math.addExact(offset, from)) % PAGE_SIZE;
+        // adjust offset and length for doing aligned page copies
+        final int adjustedFrom = from - offsetFirstPage;
+        final int adjustedLength = Math.addExact(length, offsetFirstPage);
+        // copy data using slices to ensure we do NOT materialize the pages
+        final ByteArray byteArray = BigArrays.NON_RECYCLING_INSTANCE.newByteArray(adjustedLength);
+        final BytesRefIterator iterator = slice(adjustedFrom, adjustedLength).iterator();
```

Comment:
I am not 100% about what I am about to say, but maybe this should just be simply `slice(from, length)`?
* If `from` is a large number, then `adjustedFrom` is simply the modulo with PAGE_SIZE, rather than the big number of `from` we may want (adjusted to the page boundary). Right? Also .iterator() seems to take care of the modulo inside it.
* I also wonder if `adjustedLength` should not be used for the slice. The last slice might indeed end at `length` and not have more bytes for filling the last page. But I see that .iterator() either way rounds up to PAGE_SIZE each time, so I think it's OK. Probably replacing `adjustedLength` with `length` would be equivalent.

Are there tests for the PagedBytesReference.copy() here?

Reply:
> Are there tests for the PagedBytesReference.copy() here?

AbstractBytesReferenceTestCase#testCopy covers all implementations

Reply:
Thanks! I hope that tests covers also `from` from later pages, right? I understand you mean the invocation is then OK. Apart from this, the PR LGTM. But I am now reserved to approve it because of the discussions on the main PR's page.

Reply:
I think it does, if I change the adjusted values from the original ones I can see failures.